### ISSUE TITLE
How are the words of the word classes chosen?

### ISSUE THREAD

**Issue by amirmohammadkz** [MilaNLProc/contextualized-topic-models#109]
* Contextualized Topic Models version: 2.2.1
* Python version: 3.9
* Operating System: Windows 10

### Description

I am trying to understand the meaning of each topic by analyzing the documents that have the most softmax output for each topic.

### What I Did
I applied 4 topics for CTM with `all-mpnet-base-v2` model and used `ctm.get_thetas` to get the softmax output for each topic. Then I made a Dataframe based on those results, and merged that with the documents and added a topic label which is basically the argmax of the softmax output (I assigned the topic label by choosing the topic that has the maximum softmax output)
 I assumed that word class of each topic contains the words that mostly occored in that topic. However, I have a topic with word class `exhaustion_turno_visits_incredibly_dirt` but when I count the number of `exhausion`s in the documents, there is not any document with `exhaustion_turno_visits_incredibly_dirt` label:
![image](https://user-images.githubusercontent.com/16422611/161666803-f07f62cd-902b-4d12-bfa1-ed41c912f56c.png)

I was wondering how are the word classes generated? Is it possible that have a word in word class that has not appeared in any of its documents? If not, maybe I did something wrong in assigning the labels..

**Comment by vinid** [comment]
The topic words come from the unpreprocessed documents you pass to the CTM model (either ZeroShotTM or CombinedTM)

It should never happen that you have a topic word that does not appear in the documents. 

If you can share something I am happy to take a look!

**Comment by amirmohammadkz** [comment]
Thanks for the reply!
Yeah, I understand what you are saying. Let me clarify my question

Suppose that we have two topics `exhaustion_turno_visits_incredibly_dirt`  and `today_time_got_good_day` (those words are the output of word class).

I used the training dataset for the prediction too. In the output, when I use get_thetas(), I will get two float numbers between 0 to 1 per sample. For simplicity, suppose that I just have two samples:

![image](https://user-images.githubusercontent.com/16422611/162011236-163e9314-08df-4cb6-8fdf-c749c10966fd.png)

The problem I have is that though `exhaustion` is suggested as the first keyword in the first topic, all the samples I see in the dataset has more score for the second topic, and hence, I will label them as the second topic. First, I thought maybe assigned the label mistakenly, but the second word which was `turno` only appeared in samples that has higher first topic score. 

So, my problem is that, although `exhaustion` appears in the documents, in all the samples, it actually belongs to the topic that does not have `exhaustion` in its keywords. Is it possible?

I know that words such as `good` and `today` also played a significant role in assigning the topic scores, but if this is the case, why `exhaustion` is suggested as a keyword for the first topic?

**Comment by amirmohammadkz** [comment]
Did another test with 2 topics and got another result:
[github.csv](https://github.com/MilaNLProc/contextualized-topic-models/files/8430129/github.csv)
![image](https://user-images.githubusercontent.com/16422611/162060001-86e9a37d-8969-40c4-9154-1f9caa1d61d9.png)
In this example, I just made a dataset containing two sentences. Then applied Topic modeling. As expected, all of the same sentences are categorized into the same topic, but again, we have words in the word class that did not appear in that sentence.

**Comment by vinid** [comment]
Hello :) 

Given a document, the topic predicted by the model can contain words that do not occur in the document. This is also why the ZeroShotTM works in multiple languages (e.g., it maps an Italian document to an English set of words).

In general, the document is going to be associated with the topic to which it fits most, but this does not mean that all the words in the topic will be present in the document.

Does this answer your question?

**Comment by amirmohammadkz** [comment]
Hi Federico,

Thanks for the answer :)
I understand what you are saying. But my question is regarding all the documents labeled with a specific topic, not just one. 
Just to summarize with my example:

1. I have a dataset with some documents containing the word `exhaustion`
2. I do topic modeling with 2 topics
3. The model gives me two topics. The first word in the word class of Topic 1 is `exhaustion`
4. I get softmax output of all documents with `get_thetas()`
5. All the documents containing the word `exhaustion` have softmax output <0.5 for Topic 1

Is this scenario possible?

**Comment by vinid** [comment]
My first answer would be that what you see in that case is a degenerate solution. However it also depends on the other words in the documents.

I am not sure if I have a coherent and definitive answer to this question
